server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,458 @@
+            if (isShardLimitsIncreased == false) {
+                operationTracker.getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                this.totalNodeLimitsBreachedRejections.incrementAndGet();
+            }
+            return !isShardLimitsIncreased;
+        } else {
+            boolean shardLastSuccessfulRequestDurationLimitsBreached =
+                evaluateLastSuccessfulRequestDurationLimitsBreached(operationTracker.getPerformanceTracker(), requestStartTime);
+
+            if (shardLastSuccessfulRequestDurationLimitsBreached) {
+                operationTracker.getRejectionTracker().incrementLastSuccessfulRequestLimitsBreachedRejections();
+                this.totalLastSuccessfulRequestLimitsBreachedRejections.incrementAndGet();
+                return true;
+            }
+
+            boolean shardThroughputDegradationLimitsBreached =
+                evaluateThroughputDegradationLimitsBreached(operationTracker.getPerformanceTracker(),
+                    operationTracker.getStatsTracker(), primaryAndCoordinatingThroughputDegradationLimits);
+
+            if (shardThroughputDegradationLimitsBreached) {
+                operationTracker.getRejectionTracker().incrementThroughputDegradationLimitsBreachedRejections();
+                this.totalThroughputDegradationLimitsBreachedRejections.incrementAndGet();
+                return true;
+            }
+
+            boolean isShardLimitsIncreased = increaseShardLimitSupplier.getAsBoolean();
+            if (isShardLimitsIncreased == false) {
+                operationTracker.getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                this.totalNodeLimitsBreachedRejections.incrementAndGet();
+            }
+            return !isShardLimitsIncreased;
+        }
+    }
+
+    private boolean increaseShardLimits(ShardId shardId, long nodeLimit,
+                                        LongSupplier shardCurrentBytesSupplier, LongSupplier shardLimitSupplier,
+                                        ToLongFunction<ShardIndexingPressureTracker> getShardLimitFunction,
+                                        BiPredicate<Long, Long> updateShardLimitPredicate) {
+        long currentShardLimit;
+        long newShardLimit;
+        do {
+            currentShardLimit = shardLimitSupplier.getAsLong();
+            long shardCurrentBytes = shardCurrentBytesSupplier.getAsLong();
+
+            if(((double)shardCurrentBytes / currentShardLimit) > this.upperOperatingFactor) {
+                newShardLimit = (long)(shardCurrentBytes / this.optimalOperatingFactor);
+                long totalShardLimitsExceptCurrentShard = this.shardIndexingPressureStore.getShardIndexingPressureHotStore()
+                    .entrySet().stream()
+                    .filter(entry -> (shardId != entry.getKey()))
+                    .map(Map.Entry::getValue)
+                    .mapToLong(getShardLimitFunction).sum();
+
```

Comment:
This might turn out costly based on the high concurrency of incoming traffic

Reply:
These are pure computations, without any allocation or synchronisation overhead.

Reply:
Yes I meant computation overhead, even that has a cost and time complexity

Reply:
Yes, we have done Rally benchmarking for os indexing path with backpressure feature enabled. I will be sharing the results once we have the framework level constructs rolled out as well.